File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.*;

public class Main {

    static int n;
    static int[] arr;
    static int cnt;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());

        arr = new int[n];

        nQueen(0);
        System.out.println(cnt);


    }

    private static void nQueen(int col) {
        if (col == n) {
            cnt ++;
            return;
        }

        for (int i = 0; i < n; i++) {
            arr[col] = i;

            if (check(col)) {
                nQueen(col + 1);
            }
        }
    }

    private static boolean check(int col) {
        for (int i = 0; i < col; i++) {
            if (arr[col] == arr[i]) {
                return false;
            }

            else if (Math.abs(col - i) == Math.abs(arr[col] - arr[i])) {
                return false;
            }
        }
        return true;
    }
}

File: 백준/Gold/9663. N－Queen/N－Queen.kt
import kotlin.math.abs

fun main() = with(System.`in`.bufferedReader()) {
    val n = readLine()!!.toInt()
    var cnt = 0
    var arr = IntArray(n)

    fun check(col: Int): Boolean {
        for (i in 0 until col) {
            if (arr[col] == arr[i]) {
                return false
            }
            else if (abs(col - i) == abs(arr[col] - arr[i])) {
                return false
            }
        }
        return true
    }

    fun nQueen(col: Int) {
        if (col == n) {
            cnt ++
            return;
        }

        for (i in 0 until n) {
            arr[col] = i
            if (check(col)) {
                nQueen(col + 1)
            }
        }
    }

    nQueen(0)
    println(cnt)
}